vignettes/s3-vector.Rmd

```diff
@@ -940,27 +931,223 @@ vec_arith.vctrs_meter.MISSING <- function(op, x, y, ...) {
+
+```{r}
+# for compatibility with the S4 system
+methods::setOldClass(c("pizza_percent", "vctrs_vctr"))
+```
+
+Because we've used a function from the methods package, you'll also need to add methods to `Imports`, with (e.g.) `usethis::use_package("methods")`. This is a "free" dependency because methods is bundled with every R install.
+
+Next we implement, export, and document a user-friendly helper: `percent()`.
+
+```{r}
+#' `percent` vector
+#'
+#' This creates a double vector that represents percentages so when it is
+#' printed, it is multiplied by 100 and suffixed with `%`.
+#'
+#' @param x A numeric vector
+#' @return An S3 vector of class `pizza_percent`.
+#' @export
+#' @examples
+#' percent(c(0.25, 0.5, 0.75))
+percent <- function(x = double()) {
+  x <- vec_cast(x, double())
+  new_percent(x)
+}
+```
+
+(Again note that the package name will appear in the class, but does not need to occur in the function, because we can already do `pizza::percent()`; it would be redundant to have `pizza::pizza_percent()`.)
 
-Defining S3 methods interactively is fine for iteration and exploration, but if your vector lives in a package, you also need to register the S3 methods by listing them in the `NAMESPACE` file. The namespace declarations are a little tricky because (e.g.) `vec_cast.vctrs_percent()` is both a generic function (which must be exported with `export()`) and an S3 method (which must be registered with `S3method()`). 
+### Other helpers
+
+It's a good idea to provide a function that tests if an object is of this class. If you do so, it makes sense to document it with the user-friendly constructor `percent()`:
+
+```{r}
+#' @export
+#' @rdname percent
+is_percent <- function(x) {
+  inherits(x, "pizza_percent")
+}
+```
 
-This problem wasn't considered in the design of roxygen2, so you have to be quite explicit:
+You'll also need to update the `percent()` documentation to reflect that `x` now means two different things:
 
 ```{r}
-#' @method vec_cast vctrs_percent
+#' @param x 
+#'  * For `percent()`: A numeric vector
+#'  * For `is_percent()`: An object to test.
+```
+
+Next we provide the key methods to make printing work. These are S3 methods, so they don't need to be documented, but they do need to be exported.
+
+```{r eval = FALSE}
+#' @export
+format.pizza_percent <- function(x, ...) {
+  out <- formatC(signif(vec_data(x) * 100, 3))
+  out[is.na(x)] <- NA
+  out[!is.na(x)] <- paste0(out[!is.na(x)], "%")
+  out
+}
+
 #' @export
-#' @export vec_cast.vctrs_percent
-vec_cast.vctrs_percent <- function(x, to, ...) {
-} 
+vec_ptype_abbr.pizza_percent <- function(x, ...) {
+  "prcnt"
+}
+```
+
+### Double dispatch methods for coercion and casting
+
+Next we implement methods for `vec_ptype2()` and `vec_cast()`, along with all the roxygen2 boilerplate needed to make them work. `vec_ptype2()` and `vec_cast()` have exactly the same structure, so we'll explain everything via `vec_ptype2()` and just show the result for `vec_cast()`.
```

Comment:
> have exactly the same structure

Unfortunately this isn't true. ptype2 dispatches first on the first argument, whereas cast dispatches first on the second argument.